napari/layers/_scalar_field/scalar_field.py

```diff
@@ -588,6 +590,106 @@ def _get_value(self, position):
 
         return value
 
+    def _get_value_ray(
+        self,
+        start_point: Optional[np.ndarray],
+        end_point: Optional[np.ndarray],
+        dims_displayed: list[int],
+    ) -> Optional[int]:
+        """Get the first non-background value encountered along a ray.
+
+        Parameters
+        ----------
+        start_point : np.ndarray
+            (n,) array containing the start point of the ray in data coordinates.
+        end_point : np.ndarray
+            (n,) array containing the end point of the ray in data coordinates.
+        dims_displayed : List[int]
+            The indices of the dimensions currently displayed in the viewer.
+
+        Returns
+        -------
+        value : Optional[int]
+            The first non-background value encountered along the ray. If none
+            was encountered or the viewer is in 2D mode, returns None.
+        """
+        if start_point is None or end_point is None:
+            return None
+        if len(dims_displayed) == 3:
+            # only use get_value_ray on 3D for now
+            # we use dims_displayed because the image slice
+            # has its dimensions  in th same order as the vispy
+            # Volume
+            # Account for downsampling in the case of multiscale
+            # -1 means lowest resolution here.
+            start_point = (
+                start_point[dims_displayed]
+                / self.downsample_factors[-1][dims_displayed]
+            )
+            end_point = (
+                end_point[dims_displayed]
+                / self.downsample_factors[-1][dims_displayed]
+            )
+            start_point = cast(np.ndarray, start_point)
+            end_point = cast(np.ndarray, end_point)
+            sample_ray = end_point - start_point
+            length_sample_vector = np.linalg.norm(sample_ray)
+            n_points = int(2 * length_sample_vector)
+            sample_points = np.linspace(
+                start_point, end_point, n_points, endpoint=True
+            )
+            im_slice = self._slice.image.raw
+            # ensure the bounding box is for the proper multiscale level
+            bounding_box = self._display_bounding_box_at_level(
+                dims_displayed, self.data_level
+            )
+            # the display bounding box is returned as a closed interval
+            # (i.e. the endpoint is included) by the method, but we need
+            # open intervals in the code that follows, so we add 1.
+            bounding_box[:, 1] += 1
+
+            clamped = clamp_point_to_bounding_box(
+                sample_points,
+                bounding_box,
+            ).astype(int)
```

Comment:
Why do we need this?

Reply:
I tried to look back into #3037; I don't really understand why this is needed, given that `start_point` and `end_point` should already be correctly computed here:

https://github.com/napari/napari/blob/c2c929a518bfef65d0836a723e6df61e48a54639/napari/layers/base/base.py#L1393-L1398

@kevinyamauchi @alisterburt, any thoughts?

Reply:
I expect that there might be some floating point arithmetic errors that you can avoid by clamping. (ie start/end/linspace end up being epsilon outside the volume and mucking everything up).

We could try removing it and seeing what breaks but I agree with @brisvag, let's do it in a follow-up PR.